### ISSUE TITLE
Presolve expression with same domain and orde.

### ISSUE THREAD

**Issue by yasirroni** [cvxpy/cvxpy#2550]
**Is your feature request related to a problem? Please describe.**
Assume that a f1 return 2 x**2 while f2 return x ** 2. If user want to use both f1 and f2 as objective with f1 - f2, cvxpy will fails.

```
import cvxpy as cp

v = cp.Variable(1)

# example a convex problem that cvxpy fails
f1 = 2 * cp.power(v, 2)  # convex
f2 = 1 * cp.power(v, 2)  # convex

f = f1 - f2  # equal to cp.power(v, 2) # convex - convex with same power ** 2, resulted in convex but detected as unknown

objective = cp.Minimize(cp.sum(f))

prob = cp.Problem(objective)

sol = prob.solve()
print(f'{sol=}')
print(f'{prob.status=}')
print(f'{v.value=}')
print(f'{prob.solver_stats=}')
```

**Describe the solution you'd like**
When getting current DCP error, check if both convex power have same domain, and if same domain and same method with same parameters, try to return new expression that is valid convex.

**Additional context**
Another approach is allow user to hardcode that f can be treated as convex, like `quad_form` `assume_PSD` parameter.

**Comment by Transurgeon** [comment]
Hey @yasirroni I don't think we plan on adding support for a convex wrapper. 
You can see the source code (and an example) here: https://github.com/cvxpy/cvxpy/blob/3c37fd7d1de1b07f5c73247c8f21795bb50d1c5f/cvxpy/atoms/affine/wraps.py#L70.
Please try to create your own and let us know if it works for your use case. (You would simply need to overwrite the ``is_convex`` function to always be true similarly to the nonneg wrapper above.

**Comment by aszekMosek** [comment]
@yasirroni The DCP rules impose a certain balance. On one hand they make it fairly easy for the user to both express the model  and prove its convexity all at once (I find the part about forcing the user to prove convexity quite valuable). On the other hand they make it fairly easy for the parser to both process the model and verify its convexity. Your suggestion shifts more responsibility from the user onto the parser. Do you have a real-life application where this would be useful (not a toy example)?

**Comment by yasirroni** [comment]
> Please try to create your own and let us know if it works for your use case. (You would simply need to overwrite the is_convex function to always be true similarly to the nonneg wrapper above).

Does it means that cvxpy open PR about this?





> @yasirroni The DCP rules impose a certain balance. On one hand they make it fairly easy for the user to both express the model and prove its convexity all at once (I find the part about forcing the user to prove convexity quite valuable). On the other hand they make it fairly easy for the parser to both process the model and verify its convexity. Your suggestion shifts more responsibility from the user onto the parser. Do you have a real-life application where this would be useful (not a toy example)?

I also like to use `cvxpy` because of that. Currently, I haven't meet exact use case. My use case already solved with `quad_form`, that is `a1 * x1 ** 2 + a2 * x2 **2`, where `a1` pos and `a2` neg. This particular case happens when quadratic function of power plant have different sign on their data. But again, my use case have different `x` and can be solved used `quad_form`. The problem I posted here appears when I teach my student about convexity and domain in `cp.power()`.

**Comment by yasirroni** [comment]
> I find the part about forcing the user to prove convexity quite valuable

@aszekMosek This part is nice, and I also like it. But, if user know it is convex, and cvxpy fails to recognize it, didn't we should let user pass it to solver without cvxpy enforcing convexity?

**Comment by yasirroni** [comment]
@aszekMosek I found a use case in my field, that is power systems, especially generator cost function.

Generator cost function can be written in:

$$
cost(p) = a_{0} + a_{1} * p + a_{2} * p^{2}
$$

If we want to make a sell cost, it should be a function of production cost. Thus, at least, it became:

$$
gross(p) = w * cost(p)
$$

To calculate profit, there will be some function related to both `f` and `g`:

$$
profit(p) = gross(p) - cost(p)
$$

```python
# Define the problem
v = cp.Variable(1)

# Example a convex problem that cvxpy fails
f = 2 * cp.power(v, 2)
g = 3 * f

h = g - f  # equal to cp.power(v, 2)

# Objective function
objective = cp.Minimize(cp.sum(h))

# Define the problem
prob = cp.Problem(objective)

# Attempt to solve the problem
try:
    sol = prob.solve()
    print(f'{sol=}')
    print(f'{prob.status=}')
    print(f'{v.value=}')
    print(f'{prob.solver_stats=}')
except DCPError as e:
    print(e)
    pass
```

**Comment by aszekMosek** [comment]
@yasirroni It is an interesting example, however the way you write it creates the false impression that the model would remain convex after replacing the coefficients 2, 3 by other numbers, or even after replacing the functions ``f,g`` by some other functions, whereas in reality it depends on some other structure, namely on the convexity of ``g-f``. Normally it is the user's job to find the convex structure and communicate it clearly to cvxpy. I doubt that analyzing arbitrary algebraic expressions to detect their convexity is something cvxpy wants to do, but I am not a cvxpy core developer so maybe I'm wrong.

**Comment by yasirroni** [comment]
@aszekMosek Yeah, in the example above, g - f are convex. If cvxpy not know that it is convex or not, "presolve" here intended to resolve, at least some function that is easy (power with same input and degree), or let user give it an info that it is convex and force it to the solver (skipping cvxpy logic). The second answer is not desirable as far as I know in cvxpy community and developer.
